### ISSUE TITLE
[Bug]: The stock variable is not working

### ISSUE THREAD

**Issue by pitsi** [brndnmtthws/conky#1897]
### What happened?

I am on debian testing/unstable x64 and I got the upgrade to 1.20.2 yesterday (from 1.19.6). And although I have a very specific conky layout for more than a decade now, I check for variables that may have changed and for new ones. And I came accross the stock variable, so I wanted to try it and I just put it on my .conkyrc (seen below).
https://conky.cc/variables#stock

As it seems, it does not work, because when launcing conky from the terminal, curl tries again and again to parse something that is not there. I have no idea about programming, but I had a look at the source code and it seems this is the url it tries to parse. 
`http://download.finance.yahoo.com/d/quotes.csv`

And it just can not open on my end. I do not know if it is geoblocked, but I asked a friend to test it with a vpn and a US ip, but it still fails.

p.s. Can someone please inform me if the weather variables have been removed? I do not want to open a new issue just for it.

### Version

1.20.2

### Which OS/distro are you seeing the problem on?

Debian

### Conky config

```Lua
conky.config = {
    alignment = 'bottom_right',
    background = true,
    border_width = 0,
    cpu_avg_samples = 2,
    default_color = 'ghost white',
    default_outline_color = 'black',
    default_shade_color = 'black',
    double_buffer = true,
    draw_borders = false,
    draw_graph_borders = true,
    draw_outline = false,
    draw_shades = false,
    extra_newline = false,
    font = 'ITC Avant Garde Gothic Pro:size=16',
    font1 = 'ITC Avant Garde Gothic Pro:size=14',
    gap_x = 0,
    gap_y = 32,
    minimum_height = 5,
    minimum_width = 5,
    net_avg_samples = 2,
    no_buffers = true,
    out_to_console = false,
    out_to_ncurses = false,
    out_to_stderr = false,
    out_to_x = true,
    own_window = true,
    own_window_class = 'Conky',
    own_window_type = 'desktop',
    own_window_transparent = true,
    own_window_hints = 'undecorated,below,sticky,skip_taskbar,skip_pager',
    show_graph_range = false,
    show_graph_scale = false,
    stippled_borders = 0,
    update_interval = 1.0,
    uppercase = false,
    use_spacer = 'none',
    use_xft = true,
}

--[[
# ${rss http://rss.accuweather.com/rss/liveweather_rss.asp?metric=1&locCode=(removed) 300 item_title 0} 
# ${rss https://weather-broker-cdn.api.bbci.co.uk/en/observation/rss/(removed) 300 item_title 0}
# ${rss https://weather-broker-cdn.api.bbci.co.uk/en/forecast/rss/3day/(removed) 300 item_title 0} 
# ${freq_g 1}G 
# ${curl https://wttr.in/(removed)?format=3}
]]


conky.text = [[
$alignr ${stock AAPL name}
$alignr ati ${hwmon 0 temp 1 1 0}C 
$alignr sdb ${hwmon 4 temp 1 1 0}C 
$alignr cores ${hwmon 1 temp 2 1 -15}C ${freq_g 1}G ${hwmon 1 temp 3 1 -15}C ${freq_g 2}G 
$alignr ${fs_bar 5,300 /} 
$alignr ${rss https://weather-broker-cdn.api.bbci.co.uk/en/observation/rss/(removed) 300 item_title 0} 
$alignr $font1 ${rss https://weather-broker-cdn.api.bbci.co.uk/en/observation/rss/(removed) 300 item_desc 0} 
]]
```

### Relevant log output

```Shell
$ conky 
conky: desktop window (0x53a) is root window
conky: window type - desktop
conky: drawing to created window (0xe00001)
conky: drawing to double buffer
conky: forked to background, pid is 13050
conky: FOUND: console
conky: FOUND: ncurses
conky: FOUND: file
conky: FOUND: x11
conky: FOUND: wayland
conky: 'openbox' x11 session running '(null)' destop
conky: curl: could not retrieve data from server
conky: curl: could not retrieve data from server
conky: curl: could not retrieve data from server
```

**Comment by lineage-of-roots** [comment]
@pitsi 

Google says `http://download.finance.yahoo.com/d/quotes.csv` has not worked since 2017...
So that code in there is old, and the info on conky.cc is out dated as well.

A kind Gentleman on the internet I found  pointed out the new url and format which goes like this

 `https://query1.finance.yahoo.com/v7/finance/download/^GSPC?period1=1459468800&period2=1494028800&interval=1d&events=history&crumb=yt6hjEL5Tx8 `

That will get csv files, but I currently don't know what the period interval numbers mean. 

I will see what I can do about the Conky code and also try to figure out  how the period numbers correspond to dates.

**Comment by pitsi** [comment]
Yes, when you see timestamps that are more than a billion, they are on unix epoch time, which starts on that date, counts in seconds and you can get the current one with `date +%s`.

As for the code being old, just imagine that until less than a year ago, I was checking conky's variables and configuration settings on these pages on sourceforge! That is why I asked if the weather variable stil exists.
https://conky.sourceforge.net/variables.html
https://conky.sourceforge.net/config_settings.html

**Comment by lineage-of-roots** [comment]
Oh I knew about the epoch time. 
It's just that the url the confused me.
I thought it was some fancy finance thing that accountants understood lol. (I know nothing about stocks or finance)

As for the weather variable, I am not currently seeing it anywhere in the source code. I have the latest git clone.

I did manage to get conky to retrieve the csv file from Yahoo though. The rest of the code works, just had to change the url a bit.
However, it's ugly looking right now  :)

And from what I am seeing, its always the same kind of data. Can only change the time period and  interval like daily or weekly.  Can't seem to get the fancier stuff.
Maybe it's my lack of knowledge about this stuff...Or perhaps Yahoo only wants to show it on their website now.

**Comment by Caellian** [comment]
> https://query1.finance.yahoo.com/v7/finance/download/^GSPC?period1=1459468800&period2=1494028800&interval=1d&events=history&crumb=yt6hjEL5Tx8

This now returns error along the lines of "user not logged in".

Anyway, I was looking into this as part of a separate thing and for some reason, raw currency conversion JSON data  is apparently something a lot of people really feel like it should be behind a paywall for some reason... And then they're like "oh, but it's free and can be scraped easily" and then they push in a bunch of useless features (for our use case) in order to turn it into a product...

I don't think Yahoo is a good choice for this because it changes several times per year which is annoying to maintain.

I found a nice [free API here](https://github.com/fawazahmed0/exchange-api) that is a good replacement, but the issue is that it's only updated daily, which is a regression from previous functionality.

**Comment by pitsi** [comment]
Correct me if I am wrong, but...
The forementioned apis show exchange rates for currencies, not stocks.

**Comment by Caellian** [comment]
They did both. But the state of things isn't much better for stocks either. I spent a solid morning (~5 hrs) looking into this and the basic gist of it is that all this data is sold and "free" options are request limited.

Yahoo shows these details for free, but they themselves buy the data and then show it on the website. There is no free API to access it and the only option is to write a selenium scraper that violates the Yahoo ToS. The data is loaded dynamically into the websites from a web worker with a bunch of obfuscated JS.

[`thirtyfour`](https://github.com/Vrtgs/thirtyfour) (selenium) works well in Rust and we could bundle it into conky, however it still takes about a second to fetch a single stock price. Then there's the question of whether users actually want conky to spin up a browser instance every update, or run one in the background (and check whether stock variables are used).

Other than that, bindings for paid APIs can be added to conky and users can provide their own API keys. But I was really looking for a 0-cost solution. For instance [Alphavantage](https://www.alphavantage.co/) has 25 requests/day free tier but their paid tiers are very expensive, some only offer end-of-day data on free tier, etc.
If we go that route, it would be wise to create some common structs that are populated by these various APIs.

In any case, fixing this feels like a large undertaking.

**Comment by pitsi** [comment]
Well, if there is no api to use, I think it would be better to remove the functionality and call it a day :(

**Comment by Caellian** [comment]
Yeah, for now it would be good to remove the variable as its broken. Adding a note next to it on the website is also important so people who used it before can figure out that it was removed and why.

**Comment by pitsi** [comment]
Please bot, do not close my issue. At least not until next release :)

**Comment by Caellian** [comment]
Will be removed FYI. I'm not fully done testing #2414 yet, but that PR removes the stock variable because there's nothing that can replace it. There isn't even an "48h stale data" option out there last I checked, so we'd have to maintain the stock information with custom scrapers ourselves and that's not a very attractive solution (breaks all the time, maintenance burden).

Dunno, if you care about having live stock data, the best solution is probably using such paid HTTP API from Lua.

I'll close this now as wontfix because the PR will close it with wrong status later. Sorry :/
